BlueprintUI/Sources/Layout/GridRow.swift

```diff
@@ -0,0 +1,244 @@
+    enum Width: Equatable {
+        /// Assign the child a fixed width equal to the payload.
+        case absolute(CGFloat)
+        /// Assign the child a proportional width of the available layout width. Note that proportional children
+        /// take proportional shares of the available layout width.
+        ///
+        /// ## Example:
+        ///     Layout width: 100
+        ///     Child A: .proportional(1)  -> 25 (100 * 1/4)
+        ///     Child B: .proportional(3) -> 75 (100 * 3/4)
+        ///
+        /// ## Example:
+        ///     Layout width: 100
+        ///     Child A: .proportional(0.25)  -> 25 (100 * 1/4)
+        ///     Child B: .proportional(0.75) -> 75 (100 * 3/4)
+        case proportional(CGFloat)
+    }
+}
+
+// MARK: - layout -
+extension GridRow {
+    struct GridRowLayout: Layout {
+        let verticalAlignment: VerticalAlignment
+        let spacing: CGFloat
+
+        static var defaultTraits: GridRow.Width {
+            .absolute(0)
+        }
+
+        func measure(in constraint: SizeConstraint, items: [(traits: Width, content: Measurable)]) -> CGSize {
+            guard items.count > 0 else {
+                return .zero
+            }
+
+            // Measure the the row to fit its tallest child.
+            let height = _frames(in: constraint, items: items).lazy.map { $0.height }.max() ?? 0
+
+            return CGSize(width: constraint.width.maximum, height: height)
+        }
+
+        func layout(size: CGSize, items: [(traits: Width, content: Measurable)]) -> [LayoutAttributes] {
+            guard items.count > 0 else {
+                return []
+            }
+
+            return _frames(in: SizeConstraint(size), isExactConstraint: true, items: items)
+                .map(LayoutAttributes.init)
+        }
+
+        /// Compute child frames in a given layout space.
+        ///
+        /// First, we measure the size of absolutely-sized children. Next, we measure the size of
+        /// proportionally-sized children, allocating them a proportion of the layout width available after spacing
+        /// and absolutely-sized children are subtracted.
+        ///
+        /// Finally, we compute the frames of children within the row from left to right, assigning a y-origin
+        /// and height which fulfill the row's vertical alignment.
+        ///
+        /// - Parameters:
+        ///   - constraint: A constraint describing the available layout space.
+        ///   - isExactConstraint: Whether or not the constraint describes the row's size exactly. During measurement,
+        ///   it does not. During layout, it does.
+        ///   - items: Children to layout.
+        private func _frames(
+            in constraint: SizeConstraint,
+            isExactConstraint: Bool = false,
+            items: [(traits: Width, content: Measurable)]
+        ) -> [CGRect] {
+            var sizes: [CGSize] = (0..<items.count).map { _ in .zero }
```

Comment:
`Array(repeating: CGSize.zero, count: items.count)` _might_ be slightly better for this?